multi_camera_parallel_grabber/ex-multi_camera_parallel_grabber.cpp

```diff
@@ -0,0 +1,266 @@
+#include <ifm3d/fg.h>
+#include <ifm3d/image.h>
+#include <opencv2/core/core.hpp>
+#include <opencv2/imgcodecs.hpp>
+#include <opencv2/highgui/highgui.hpp>
+
+namespace
+{   
+  // enum for all the trigger types supported by o3d3xx device
+  enum class trigger_mode : int { FREE_RUN = 1, SW = 2, POSITIVE_EDGE = 3, NEGATIVE_EDGE = 4, POSITIVE_N_NEGATIVE = 5 };
+
+  //flag for stoping the application
+  std::atomic<bool> start = true;
+
+  //Add the IP of cameras to be used for the complete stystem. Address at 0th location must be of camera with software trigger 
+  /* to add device add device address at second last position in the camera_ips vector, increase NUMBER_OF_DEVICES too */
+  const auto NUMBER_OF_DEVICES = 3;
+  const std::array<std::string, NUMBER_OF_DEVICES> camera_ips = { "192.168.0.70", "192.168.0.71","192.168.0.72"};
+
+  auto j_conf = json::parse(R"(
+        {
+          "ifm3d":
+          {
+            "Device":
+            {
+              "ActiveApplication": "1"
+            },
+            "Apps":
+            [
+              {
+                "Index": "1",
+                "TriggerMode":"2",
+                "LogicGraph": "{\"IOMap\": {\"OUT1\": \"RFT\",\"OUT2\": \"AQUFIN\"},\"blocks\": {\"B00001\": {\"pos\": {\"x\": 200,\"y\": 200},\"properties\": {},\"type\": \"PIN_EVENT_IMAGE_ACQUISITION_FINISHED\"},\"B00002\": {\"pos\": {\"x\": 200,\"y\": 75},\"properties\": {},\"type\": \"PIN_EVENT_READY_FOR_TRIGGER\"},\"B00003\": {\"pos\": {\"x\": 600,\"y\": 75},\"properties\": {\"pulse_duration\": 0},\"type\": \"DIGITAL_OUT1\"},\"B00005\": {\"pos\": {\"x\": 600,\"y\": 200},\"properties\": {\"pulse_duration\": 0},\"type\": \"DIGITAL_OUT2\"}},\"connectors\": {\"C00000\": {\"dst\": \"B00003\",\"dstEP\": 0,\"src\": \"B00002\",\"srcEP\": 0},\"C00001\": {\"dst\": \"B00005\",\"dstEP\": 0,\"src\": \"B00001\",\"srcEP\": 0}}}",
+                "Imager":
+                {
+                    "ExposureTime": "1000",
+                    "Type":"under5m_moderate",
+                    "FrameRate":"20"
+                }
+              }
+            ]
+          }
+        }
+    )");
+
+  //for configuration of required trigger on camera
+  void configuration(ifm3d::Camera::Ptr& camera, trigger_mode type)
+  {
+      try
+      {
+          auto application_id = camera->ActiveApplication();
+          j_conf["ifm3d"]["Device"]["ActiveApplication"] = std::to_string(application_id);
+          j_conf["ifm3d"]["Apps"][0]["Index"] = std::to_string(application_id);
+          j_conf["ifm3d"]["Apps"][0]["TriggerMode"] = std::to_string((int)type);
+
+          camera->FromJSON(j_conf);
+      }
+      catch (std::exception &e)
+      {
+          std::cout << e.what();
+      }
+  }
+
+  // class to encapsulate all the camera related stuff and grabbing images 
+  struct  CameraObject
+  {
+    using Ptr = std::shared_ptr<CameraObject>;
+    using Callback = std::function<void(ifm3d::ImageBuffer::Ptr &image_buffer, bool in_time)>;
```

Comment:
same comment as above -- naked pointers better express semantics of ownership here. `ifm3d::ImageBuffer* pImageBuffer` instead of a reference shared_ptr.